Fraction numerator denominator helpers

AnslemHack opened the pull request:
This PR fixes #3595  introduces two new arithmetic functions, num() and den(), which extract the numerator and denominator from Fraction objects respectively.

gwhitney commented on src/function/arithmetic/den.js:
```diff
@@ -0,0 +1,39 @@
+  ({ typed }) => {
+    /**
+     * Get the denominator of a fraction.
+     * For a fraction `a/b`, the function returns `b`.
```
I agree with Jos; you may just want to note that the denominator is always positive. Also the same comments about lowest terms go here.

gwhitney commented:
A few more items:

* `re()` and `im()` are in src/function/complex. Shouldn't these be in a new subdirectory src/function/fraction? (Jos?)
* Please add some tests showing that num and den do reasonable things on number and bigint input.
* Presumably at the moment num and den will fail on BigNumber input, because there is no implicit conversion from BigNumber to Fraction. But it seems to me that for these two particular functions, what else could one mean but to convert the BigNumber to Fraction and then return the numerator or denominator? So, with @josdejong's blessing, please add an explicit signature for BigNumber for both functions that does the conversion and then returns the numerator or denominator. And then tests for this behavior, of course. Thanks!

AnslemHack commented:
> A few more items:
> 
> * `re()` and `im()` are in src/function/complex. Shouldn't these be in a new subdirectory src/function/fraction? (Jos?)
> * Please add some tests showing that num and den do reasonable things on number and bigint input.
> * Presumably at the moment num and den will fail on BigNumber input, because there is no implicit conversion from BigNumber to Fraction. But it seems to me that for these two particular functions, what else could one mean but to convert the BigNumber to Fraction and then return the numerator or denominator? So, with @josdejong's blessing, please add an explicit signature for BigNumber for both functions that does the conversion and then returns the numerator or denominator. And then tests for this behavior, of course. Thanks!

thanks for your feedback I have now addressed your concerns. please take another look.

gwhitney commented on test/typescript-tests/testTypes.ts:
```diff
@@ -3181,3 +3181,63 @@ Match types of exact positional arguments.
+  const fractionMatrix = math.matrix([math.fraction(1, 2), math.fraction(3, 4)])
+  expectTypeOf(math.num(fractionMatrix)).toMatchTypeOf<Matrix>()
+  expectTypeOf(math.den(fractionMatrix)).toMatchTypeOf<Matrix>()
+  assert.deepStrictEqual(
```
Great, thanks! Just need to also put the `<bigint>` into `.toMatchTypeOf<MathArray>()` in a couple of places above as well.

gwhitney replied:
Looks good, thanks.

gwhitney commented:
Yes, very close now. Just a couple of open items, and please in addition to taking care of those also update to be on top of the latest version of the `develop` branch, thanks. Should be ready to merge after that.

gwhitney commented on src/function/fraction/den.js:
```diff
@@ -0,0 +1,48 @@
+    return typed(name, {
+      Fraction: (x) => x.d,
+      BigNumber: (x) => fraction(x).d,
+      'Array | Matrix': typed.referToSelf((self) => (x) => deepMap(x, self))
```
Fixed now, thanks.
